draft-specifications/CapTP Specification.md

```diff
@@ -0,0 +1,978 @@
+[Cryptography](#cryptography)
+
+### Receiving
+
+The `captp-version` MUST be equal to `1.0`. If the version does not match, the
+connection MUST be aborted.
+
+The `acceptable-location-sig` MUST be valid that the `session-pubkey` provided a
+valid signature of `acceptable-location`.
+
+## [`op:bootstrap`](#op-bootstrap)
+
+This operation requests the bootstrap object to be available at the specified
+`answer-position`. The message looks like:
+
+```text
+<op:bootstrap [answer-position    ; positive integer or 0
+               resolve-me-desc]>  ; desc:import-object | desc:import-promise
+```
+
+### The bootstrap Object
+
+The bootstrap object is responsible for providing access to local objects on the
+session. It has two different behaviors, these are selected using the
+conventional CapTP method mechanism of sending a symbol as the first argument,
+the following methods are available:
+
+-   `fetch`
+-   `deposit-gift`
+-   `withdraw-gift`
+
+#### `fetch` Method
+
+This method is used to fetch an object from the bootstrap object. To use it you
+need a `swiss-number` which is a Binary Data type. This swiss number should
+correspond an object which exists in this session. The result will be the object
+which corresponds to this `swiss-number` or an error if the object does not
+exist or a swiss number was not provided.
+
+An example of how to use this method is:
+
+```text
+<op:deliver <desc:export 0>          ; Remote bootstrap object
+            ['fetch                  ; Argument 1: Symbol "fetch"
+             swiss-number]           ; Argument 2: Binary Data
+            3                        ; Answer position
+            <desc:import-object 5>>  ; object exported by us at position 5 should provide the answer
+```
+
+#### `deposit-gift` Method
+
+The deposit gift method is used in conjunction with sending a [Third Party
+Handoff](#third-party-handoffs). This method is used to deposit a gift which has
+been sent to the bootstrap object. It has two arguments:
+
+1.  A gift ID that is positive integer.
+2.  A `desc:import-object` or `desc:import-promise` which has been exported
+    within the given CapTP session.
+
+Here is an example of how to use this method:
+
+```text
+<op:deliver-only <desc:export 0>                               ; Remote bootstrap object
+                  ['deposit-gift                               ; Argument 1: Symbol "deposit-gift"
+                  gift-id                                      ; Argument 2: Positive integer or 0
+                  <desc:import-object | desc:import-promise>)> ; Argument 3
+```
+
+#### `withdraw-gift` Method
+
+This method is used to send the [`desc:handoff-receive`](#desc-handoff-receive)
+in order to receive a gift. It has one arguments:
+
+- The `desc:handoff-receive`
+
+This should have been sent with the `op:deliver` operation, the response the
+bootstrap object should give is the gift which was (or will be) deposited.
+
```

Comment:
This actually needs two arguments:
1. gift id of the gift
2. VatId of the giver